### PR TITLE
Simplify gRPC service definition for the worker

### LINKED ISSUE #6074: Clean up gRPC server abstractions in the core worker
We should clean up gRPC in the core worker as follows:
1. There should be a common gRPC core worker handler. The handler methods can either post events onto the main event loop, or call other core worker methods directly.
2. No other interfaces.

This way, a worker->worker and raylet->worker RPCs can be added with fewer changes: (1) defining the proto message, (2) writing the handler method, and (3) the client handling code.

We should also probably add a worker->raylet RPC client. I believe the only thing preventing this currently is plumbing the node manager port into the core worker.

cc @edoakes @stephanie-wang

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

It's currently complicated to add a new endpoint to the worker's gRPC server and also opaque where the endpoint handlers are run. This change makes it clear, as the `CoreWorker` handles all endpoints and posts them to the appropriate `io_service` as needed and also adds a macro so that all C++ changes happen in one place.

## Related issue number

Closes #6074

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/core_worker/core_worker.h [hunk 1]
```diff
@@ -20,6 +20,15 @@
 #include "ray/rpc/worker/worker_client.h"
 #include "ray/rpc/worker/worker_server.h"
 
+/// The set of gRPC handlers and their associated level of concurrency. If you want to
+/// add a new call to the worker gRPC server, first change src/ray/protobuf/worker.proto,
+/// then add a new RAY_CORE_WORKER_RPC_HANDLER here, then add the implementation to the
+/// CoreWorker class below.
+#define RAY_CORE_WORKER_RPC_HANDLERS                      \
+  RAY_CORE_WORKER_RPC_HANDLER(AssignTask, 5)              \
+  RAY_CORE_WORKER_RPC_HANDLER(DirectActorAssignTask, 100) \
```
Comment (ericl) [inline src/ray/core_worker/core_worker.h]: How about 9999 for this one?

File: src/ray/core_worker/core_worker.h [hunk 2]
```diff
@@ -466,6 +498,9 @@ class CoreWorker {
   // Interface that receives tasks from the raylet.
   std::unique_ptr<CoreWorkerRayletTaskReceiver> raylet_task_receiver_;
 
+  /// The rpc service for `WorkerTaskService`.
+  rpc::WorkerGrpcService grpc_service_;
```
Comment (ericl) [inline src/ray/core_worker/core_worker.h]: common rpc service for all core worker modules

File: src/ray/protobuf/worker.proto [hunk 3]
```diff
@@ -8,6 +8,30 @@ message ActiveObjectIDs {
   repeated bytes object_ids = 1;
 }
 
+// Persistent state of an ActorHandle.
```
Comment (ericl) [inline src/ray/protobuf/worker.proto]: Can you rename this file to core_worker.proto?

**Review by ericl — APPROVED** [review]
LGTM if tests pass

File: src/ray/core_worker/core_worker.h [hunk 4]
```diff
@@ -295,6 +304,29 @@ class CoreWorker {
                                const std::vector<std::shared_ptr<Buffer>> &metadatas,
                                std::vector<std::shared_ptr<RayObject>> *return_objects);
 
+  /* Handlers for the worker's gRPC server. These are executed on the io_service_ and post
```
Comment (pcmoritz) [inline src/ray/core_worker/core_worker.h]: We should avoid using `/* */` comments (it prevents you from being able to comment things out with them as they can't be nested)

File: src/ray/protobuf/worker.proto [hunk 5]
```diff
@@ -20,8 +44,50 @@ message AssignTaskRequest {
...
 }
 
-// Service for worker.
-service WorkerTaskService {
+message ReturnObject {
+  // Object ID.
+  bytes object_id = 1;
+  // Data of the object.
+  bytes data = 2;
+  // Metaata of the object.
```
Comment (pcmoritz) [inline src/ray/protobuf/worker.proto]: typo

File: src/ray/rpc/worker/worker_client.h [hunk 6]
```diff
@@ -37,19 +56,119 @@ class WorkerTaskClient {
...
+                *stub_, &WorkerService::Stub::PrepareAsyncAssignTask, request, callback);
     return call->GetStatus();
   }
 
+  /// Push a task.
+  ///
+  /// \param[in] request The request message.
+  /// \param[in] callback The callback function that handles reply.
+  /// \return if the rpc call succeeds
+  ray::Status DirectActorAssignTask(
```
Comment (stephanie-wang) [inline src/ray/rpc/worker/worker_client.h]: Can we put these methods in a .cc?

File: src/ray/core_worker/core_worker.h [hunk 7]
```diff
@@ -20,6 +20,15 @@
 #include "ray/rpc/worker/worker_client.h"
 #include "ray/rpc/worker/worker_server.h"
 
+/// The set of gRPC handlers and their associated level of concurrency. If you want to
+/// add a new call to the worker gRPC server, first change src/ray/protobuf/worker.proto,
+/// then add a new RAY_CORE_WORKER_RPC_HANDLER here, then add the implementation to the
+/// CoreWorker class below.
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.h]: Can you be more specific about the naming convention that you need to use for the corresponding code in the protobuf and CoreWorker?

**Review by stephanie-wang — COMMENTED** [review]
Thanks, this is great!

File: src/ray/rpc/worker/worker_client.h (hunk 6 above)
Comment (edoakes) [reply src/ray/rpc/worker/worker_client.h]: Good idea but I tried doing this and got a bunch of thread annotation errors (which I think means the existing thread annotations aren't actually working), so I will leave this to @ericl.

File: src/ray/core_worker/core_worker.h (hunk 7 above)
Comment (edoakes) [reply src/ray/core_worker/core_worker.h]: Done. Added an example for clarity

**Comment by ericl** [comment]
Merging to unblock direct task calls -- feel free to iterate further in parallel.

**Pull request merged**